File: backend/test_signup_flow.py
#!/usr/bin/env python3
"""
Test script to verify the complete signup flow works correctly
"""

import requests
import json
import time

# Configuration
BASE_URL = "http://localhost:8000"
FRONTEND_URL = "http://localhost:3000"


def test_registration():
    """Test user registration endpoint"""
    print("🧪 Testing registration endpoint...")

    # Test data
    user_data = {
        "email": f"testflow_{int(time.time())}@example.com",
        "password": "TestPassword123!",  # pragma: allowlist secret
        "first_name": "Flow",
        "last_name": "Test",
        "role": "barber",
    }

    try:
        response = requests.post(
            f"{BASE_URL}/api/v1/auth/register",
            json=user_data,
            headers={"Content-Type": "application/json"},
        )

        if response.status_code == 200:
            print("✅ Registration successful!")
            user_info = response.json()
            print(f"   User ID: {user_info['id']}")
            print(f"   Email: {user_info['email']}")
            print(f"   Trial Status: {user_info['subscription_status']}")
            print(f"   Trial Active: {user_info['is_trial_active']}")
            print(f"   Days Remaining: {user_info['days_remaining_in_trial']}")
            return user_data["email"], user_data["password"]
        else:
            print(f"❌ Registration failed: {response.status_code}")
            print(f"   Response: {response.text}")
            return None, None

    except Exception as e:
        print(f"❌ Registration error: {e}")
        return None, None


def test_login(email, password):
    """Test user login endpoint"""
    print("\n🧪 Testing login endpoint...")

    try:
        response = requests.post(
            f"{BASE_URL}/api/v1/auth/token",
            data={"username": email, "password": password},
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

        if response.status_code == 200:
            print("✅ Login successful!")
            token_info = response.json()
            print(f"   Access Token: {token_info['access_token'][:20]}...")
            print(f"   Token Type: {token_info['token_type']}")
            return token_info["access_token"]
        else:
            print(f"❌ Login failed: {response.status_code}")
            print(f"   Response: {response.text}")
            return None

    except Exception as e:
        print(f"❌ Login error: {e}")
        return None


def test_authenticated_endpoint(token):
    """Test accessing an authenticated endpoint"""
    print("\n🧪 Testing authenticated endpoint...")

    try:
        response = requests.get(
            f"{BASE_URL}/api/v1/auth/me", headers={"Authorization": f"Bearer {token}"}
        )

        if response.status_code == 200:
            print("✅ Authenticated access successful!")
            user_info = response.json()
            print(f"   User: {user_info['first_name']} {user_info['last_name']}")
            print(f"   Email: {user_info['email']}")
            return True
        else:
            print(f"❌ Authenticated access failed: {response.status_code}")
            print(f"   Response: {response.text}")
            return False

    except Exception as e:
        print(f"❌ Authenticated access error: {e}")
        return False


def test_frontend_status():
    """Test if frontend is running"""
    print("\n🧪 Testing frontend status...")

    try:
        response = requests.get(FRONTEND_URL, timeout=5)
        if response.status_code == 200:
            print("✅ Frontend is running!")
            return True
        else:
            print(f"❌ Frontend returned status: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Frontend is not accessible: {e}")
        return False


if __name__ == "__main__":
    print("🚀 Starting Signup Flow Test\n")

    # Test backend endpoints
    email, password = test_registration()
    if email and password:
        token = test_login(email, password)
        if token:
            test_authenticated_endpoint(token)

    # Test frontend
    test_frontend_status()

    print("\n✨ Test complete! The signup flow is working correctly.")
    print("💡 You can now test the frontend signup at: http://localhost:3000/signup")

File: frontend/test-signup-flow.js
#!/usr/bin/env node

const axios = require('axios').default;

const API_URL = 'http://localhost:3000/api/proxy/api/v1';

async function testSignupFlow() {
  console.log('🧪 Testing Signup Flow...\n');

  // Generate unique email
  const timestamp = Date.now();
  const testEmail = `test${timestamp}@example.com`;
  const testData = {
    email: testEmail,
    password: 'TestPass123@',  // pragma: allowlist secret
    first_name: 'Test',
    last_name: 'User' + timestamp,
    role: 'barber'
  };

  console.log('1️⃣ Testing Registration...');
  console.log('   Email:', testEmail);

  try {
    // Test registration
    const registerResponse = await axios.post(`${API_URL}/auth/register`, testData);
    console.log('✅ Registration successful!');
    console.log('   User ID:', registerResponse.data.id);
    console.log('   Trial active:', registerResponse.data.is_trial_active);
    console.log('   Days remaining:', registerResponse.data.days_remaining_in_trial);

    // Wait a bit for DB transaction
    await new Promise(resolve => setTimeout(resolve, 1000));

    console.log('\n2️⃣ Testing Login...');

    // Test login
    const loginData = new URLSearchParams();
    loginData.append('username', testEmail);
    loginData.append('password', 'TestPass123@');

    const loginResponse = await axios.post(`${API_URL}/auth/token`, loginData, {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      }
    });

    console.log('✅ Login successful!');
    console.log('   Token length:', loginResponse.data.access_token.length);
    console.log('   User role:', loginResponse.data.user.role);
    console.log('   Permissions:', loginResponse.data.user.permissions);

    // Test authenticated request
    console.log('\n3️⃣ Testing Authenticated Request...');
    const meResponse = await axios.get(`${API_URL}/auth/me`, {
      headers: {
        'Authorization': `Bearer ${loginResponse.data.access_token}`
      }
    });

    console.log('✅ Auth verification successful!');
    console.log('   User email:', meResponse.data.email);
    console.log('   Full name:', meResponse.data.full_name);

    console.log('\n✅ All tests passed! The signup flow is working correctly.');

  } catch (error) {
    console.error('\n❌ Test failed!');
    if (error.response) {
      console.error('   Status:', error.response.status);
      console.error('   Error:', JSON.stringify(error.response.data, null, 2));
    } else if (error.request) {
      console.error('   No response received');
      console.error('   Request:', error.request._currentUrl || error.config.url);
    } else {
      console.error('   Error:', error.message);
      console.error('   Stack:', error.stack);
    }
    process.exit(1);
  }
}

// Run the test
testSignupFlow();
